### PR TITLE
Add ParallelDims helper and --pp_degree option

### PR DESCRIPTION
**Pull request by wconstab**
Make dp_degree configurable (=-1 consumes the rest of the mesh not used by PP/SP, =1 disables, >1 specifies requested data paralell size.

Local tests showed these outputs:

with --dp_degree=-1 and pp, sp = 1
Building 1-D device mesh with ['dp'], [8]

with sp=4
Building 2-D device mesh with ['dp', 'sp'], [2, 4]

with sp=2, pp=2
Building 3-D device mesh with ['dp', 'sp', 'pp'], [2, 2, 2]

with sp=2, pp=3 (WS=8)
AssertionError: Invalid parallel dims: dp(1) * sp(2) * pp(3) != WORLD_SIZE8

with sp=2, pp=3 (WS=6)
Building 2-D device mesh with ['sp', 'pp'], [2, 3]

### CODE REVIEW

File: train.py [hunk 1]
```diff
@@ -148,13 +148,22 @@ def main(args):
...
     )
     parser.add_argument(
         "--sp_degree",
         type=int,
         default=LOCAL_WORLD_SIZE,
-        help="Sequence Parallelism degree",
+        help="Sequence Parallelism degree.  1 means disabled.",
+    )
+    parser.add_argument(
+        "--pp_degree",
```
Comment (wanchaol) [inline train.py:163]: can we add some assertion or error thrown in the `parallelize_llama.py` when `sp` or `pp` > 1 we should error out currently, as we haven't add sp or pp yet.

File: train.py [hunk 2]
```diff
@@ -148,13 +148,22 @@ def main(args):
         "--steps", type=int, default=-1, help="how many train steps to run"
     )
     parser.add_argument(
-        "--enable_sp", action="store_true", help="Whether to use Sequence Parallelism."
+        "--dp_degree",
+        type=int,
+        default=-1,
+        help="Data Parallelism degree. -1 means leftover ranks will be used (After SP/PP). 1 means disabled.",
```
Comment (wanchaol) [inline train.py:154]: I wonder when we should disable DP, it seems like we would always need DP in training, so wondering if we should check dp_degree should always > 1 when world_size > 1? if world_size == 1, then dp_degree should == 1 as it falls into a single GPU case

File: train.py (hunk 1 above)
Comment (wconstab) [reply train.py:163]: good point. will do.

File: train.py (hunk 2 above)
Comment (wconstab) [reply train.py:154]: i mean, even for just debugging, you don't think its useful to be able to disable?  (try to run PP only or SP only, or PP+SP only, for instance)

Comment (wanchaol) [reply train.py:154]: That's a good callout! yeah i think we can leave the door to disable dp for pure model level debugging

**Pull request merged**